enhancements/update/change-management-and-maintenance-schedules.md

```diff
@@ -0,0 +1,1220 @@
+
+In the HCP topology, the HostedCluster and NodePool resources are enhanced to support the change management strategies
+`MaintenanceSchedule` and `Disabled`. 
+
+#### Standalone Clusters
+
+In the Standalone topology, the ClusterVersion and MachineConfigPool resources are enhanced to support the change management strategies
+`MaintenanceSchedule` and `Disabled`. The MachineConfigPool also supports the `Manual` and `Assisted` strategies. 
+
+#### Single-node Deployments or MicroShift
+
+The ClusterVersion operator will honor the change management field just as in a standalone profile. If those profiles
+have a MachineConfigPool, material changes the node could be controlled with a change management policy
+in that resource.
+
+#### OCM Managed Profiles
+OpenShift Cluster Manager (OCM) should expose a user interface allowing users to manage their change management policy.
+Standard Fleet clusters will expose the option to configure the MaintenanceSchedule strategy - including
+only permit and exclude times.
+
+- Service Delivery will reserve the right to disable this strategy for emergency corrective actions.
+- Service Delivery should constrain permit & exclude configurations based on their internal policies. For example, customers may be forced to enable permissive windows which amount to at least 6 hours a month.
+
+### Implementation Details/Notes/Constraints
+
+#### ChangeManagement Stanza
+The change management stanza will be introduced into ClusterVersion and MachineConfigPool (for standalone profiles)
+and HostedCluster and NodePool (for HCP profiles). The structure of the stanza is:
+
+```yaml
+spec:
+  changeManagement:
+    # The active strategy for change management (unless disabled by disabledUntil).
+    strategy: <strategy-name|Disabled|null>
+    
+    # If set to true or a future date, the effective change management strategy is Disabled. Date 
+    # must be RFC3339. 
+    disabledUntil: <bool|date|null>
+    
+    # If set to true or a future date, all strategies other than Disabled are paused. Date 
+    # must be RFC3339. 
+    pausedUntil: <bool|date|null>    
+    
+    # If a strategy needs additional configuration information, it can read a 
+    # key bearing its name in the config stanza.
+    config:
+       <strategy-name>:
+         ...configuration policy for the strategy...
+```
+
+#### MaintenanceSchedule Strategy Configuration
+
+```yaml
+spec:
+  changeManagement:
+    strategy: MaintenanceSchedule
+    config:
+       maintenanceSchedule:
+        # Specifies a reoccurring permissive window. 
+        permit:  
+          # RRULEs (https://www.rfc-editor.org/rfc/rfc5545#section-3.3.10) are commonly used 
+          # for calendar management metadata. Only a subset of the RFC is supported.
+          # See "RRULE Constraints" section for details.
+          # If unset, all dates are permitted and only exclude constrains permissive windows.
+          recurrence: <rrule|null>

```

Comment:
I'd be interested in your ideas. I think it would end up needing to smell similar to be expressive enough to satisfy customers unless we wanted to go with something like CEL. Other factors which encouraged the choice:
- Initial assessment of https://github.com/teambition/rrule-go looked promising to prevent significant custom code. 
- Existing standard ensures easy to find doc and supporting tooling like https://freetools.textmagic.com/rrule-generator when customers want to build or check RRULEs.

fwiw, gcloud adopted RRULE: https://cloud.google.com/kubernetes-engine/docs/how-to/maintenance-windows-and-exclusions .

Reply:
Well, what if the recurrence looked more like a structure of the various features of RRULE. We could apply succinct validation to the different parts of the RRULE that we support, and then convert to RRULE under the hood.

```
recurrence:
  frequency: Monthly
  interval: 3
  byDay:
    position: 2
    weekday: Sunday
```
Would convert down to `FREQ=MONTHLY;BYSETPOS=1;BYDAY=MO;INTERVAL=3`

But we can apply specific validations and help text to each part of the structure, and importantly, since we aren't taking in a raw string from the end user, we can control what they can input to prevent the `UNTIL` clause for example

Reply:
Pros and cons. As a customer, I would much rather use a tool to generate an rrule and directly paste it into a CRD without having to decompose it into a structure. From an engineering perspective, I want validation to prevent them from hitting seemingly arbitrary limitations. 
Personally, I side with the copy & paste use case because I expect few will be trying to craft their own string/struct and validating / reporting issues with knobs like `UNTIL` in a string input can be achieved pretty easily.

Reply:
With the struct approach that I've suggested, users don't need to even know that we are using RRULE underneath, IMO we can expose this in a way that is clear without the cognitive load to think "oh I need a generator for this".

When do you want this to happen? Monthly
Every Month or Every X Months? Every *3* Months
When in the month do you want it to happen? On the *2nd* *Sunday*

Maybe I'm over simplying it slightly but with a well designed API, I don't think it will be hard for users to understand, vs RRULE which has keywords and special meanings for particular numeric values. We can map that complexity away if we think about the naming and values a bit.

Eg Position doesn't need to be called position, and rather than 2 as a value, we could have `First|Second|Third|Fourth|Last` as an enum rather than `{1, 2, 3, 4, -1}`

I'll try to find some time to go through RRULE again and suggest an API for this